File: src/main/java/polyglotprogramming/messaging/v4java/MessageHandlers.java
package polyglotprogramming.messaging.v4java;
import polyglotprogramming.messaging.v4java.Messages.*;

class MessageHandlers {
  public static MessageHandler<Start> startMessageHandler = new MessageHandler<Start>() {
    protected OutgoingMessage process(Start message) {
      return success("started");
    }
  };

  public static MessageHandler<DoWork1> doWork1MessageHandler = new MessageHandler<DoWork1>() {
    protected OutgoingMessage process(DoWork1 message) {
      return success("work1");
    }
  };

  public static MessageHandler<DoWork2> doWork2MessageHandler = new MessageHandler<DoWork2>() {
    protected OutgoingMessage process(DoWork2 message) {
      return failed("work2");
    }
  };

  public static MessageHandler<Stop> stopMessageHandler = new MessageHandler<Stop>() {
    protected OutgoingMessage process(Stop message) {
      return success("stopped");
    }
  };
}

File: src/main/scala/polyglotprogramming/messaging/v4/MessageHandlers.scala
package polyglotprogramming.messaging.v4
import polyglotprogramming.messaging.*

trait MessageHandler[IM <: IncomingMessage]:
  final def apply(message: IM): OutgoingMessage =
    println(s"Received message: $message")
    process(message)

  protected def process(message: IM): OutgoingMessage

  protected def success(response: String): OutgoingMessage = Succeeded(response)
  protected def failure(response: String): OutgoingMessage = Failed(response)

given MessageHandler[Start] with
  protected def process(message: Start): OutgoingMessage = success("started")

given MessageHandler[DoWork1] with
  protected def process(message: DoWork1): OutgoingMessage = success("work1")

given MessageHandler[DoWork2] with
  protected def process(message: DoWork2): OutgoingMessage = failure("work2")

given MessageHandler[Stop] with
  protected def process(message: Stop): OutgoingMessage = success("stopped")
